numcodecs/fletcher32.pyx

```diff
@@ -0,0 +1,45 @@
+import struct
+
+from numcodecs.abc import Codec
+from numcodecs.compat import ensure_contiguous_ndarray
+
+from libc.stdint cimport uint8_t, uint16_t, uint32_t
+
+cdef extern from "_fletcher.c":
+    uint32_t H5_checksum_fletcher32(const void *_data, size_t _len)
+
+
+class Fletcher32(Codec):
+    """The fletcher checksum with 16-bit words and 32-bit output
+
+    With this codec, the checksum is concatenated on the end of the data
+    bytes when encoded. At decode time, the checksum is performed on
+    the data portion and compared with the four-byte checksum, raising
+    ValueError if inconsistent.
+    """
+
+    codec_id = "fletcher32"
+
+    def encode(self, buf):
+        """Return buffer plus 4-byte fletcher checksum"""
+        buf = ensure_contiguous_ndarray(buf).ravel().view('uint8')
+        cdef const uint8_t[::1] b_ptr = buf
+        val = H5_checksum_fletcher32(&b_ptr[0], buf.nbytes)
```

Comment:
At the Cython level this could just be a function that takes a `const uint8_t[::1]` object, which would allow passing the buffer through unchanged without the need to access the address, size, etc.

Reply:
I know that - it's the kind of thing I've done a bunch of times in the fastparquet codebase. However, there's  casting and prefix increments that are hard to follow in the embedded code. I just know that my earlier simpler cython version did not get the same result.